packages/op-coreutils/src/op_coreutils/bigquery/write.py

```diff
@@ -1,150 +1,235 @@
+        expiration_days (int, optional): Partition expiration in days. Defaults to 360.
+    """
     client = init_client()
+    ensure_valid_dt(df)
 
     destination = f"{dataset}.{table_name}"
 
     if df["dt"].dtype == pl.String:
         df = df.with_columns(dt=pl.col("dt").str.strptime(pl.Datetime, "%Y-%m-%d"))
 
     partitions = df["dt"].unique().sort().to_list()
-    if isinstance(client, MagicMock):
-        operation = "DRYRUN "
-    else:
-        operation = ""
     log.info(
-        f"{operation}Writing {len(partitions)} partitions to BQ [{partitions[0]} ... {partitions[-1]}]"
+        f"Writing {len(partitions)} partitions to BQ [{partitions[0]} ... {partitions[-1]}]"
     )
 
-    with io.BytesIO() as stream:
-        df.write_parquet(stream)
-        filesize = stream.tell()
-        stream.seek(0)
-        job = client.load_table_from_file(
-            stream,
-            destination=destination,
-            job_config=bigquery.LoadJobConfig(
-                source_format=bigquery.SourceFormat.PARQUET,
-                write_disposition=bigquery.WriteDisposition.WRITE_TRUNCATE,
-                time_partitioning=bigquery.TimePartitioning(
-                    type_=bigquery.TimePartitioningType.DAY,
-                    field="dt",  # Name of the column to use for partitioning.
-                    expiration_ms=expiration_days * 24 * 3600 * 1000,
-                ),
-            ),
-        )
-        job.result()
+    time_partitioning = bigquery.TimePartitioning(
+        type_=bigquery.TimePartitioningType.DAY,
+        field="dt",
+        expiration_ms=expiration_days * 24 * 3600 * 1000,
+    )
 
-        if isinstance(client, MagicMock):
-            operation = "DRYRUN "
-        else:
-            operation = ""
+    write_df_to_bq(
+        df,
+        destination,
+        client,
+        operation="OVERWRITE PARTITION",
+        time_partitioning=time_partitioning,
+    )
 
-        log.info(
-            f"{operation}OVERWRITE PARTITION: Wrote {human_rows(len(df))} {human_size(filesize)} to BQ {destination}"
-        )
+
+def upsert_partition(
+    df: pl.DataFrame, dt: str, dataset: str, table_name: str, unique_keys=["dt"]
+):
+    """Upsert data into a specific partition in a BigQuery table.
+
+    Args:
+        df (pl.DataFrame): The DataFrame to upsert.
+        dt (str): The partition date in 'YYYY-MM-DD' format.
+        dataset (str): The BigQuery dataset name.
+        table_name (str): The BigQuery table name.
+        unique_keys (list, optional): Columns that uniquely identify rows. Defaults to ['dt'].
+    """
+    client = init_client()
+    ensure_valid_dt(df, dt)
+    df = df.with_columns(dt=pl.lit(dt).str.strptime(pl.Datetime, "%Y-%m-%d"))
+
+    destination = f"{dataset}.{table_name}"
+    staging_table_name = f"{table_name}_staging"
+    staging_destination = f"{dataset}.{staging_table_name}"
+
+    write_df_to_bq(df, staging_destination, client, operation="WRITE STAGING TABLE")
```

Comment:
One note - I got errors in the past when writing to an existing `_staging` table (the table existed if a prior run errored out in the middle).

So I wonder if this needs a step to first delete the staging table if it exists?

Reply:
Great point! To ensure the function handles this, I will add step to delete the staging table if it exists already.